docs(firestore): added pagination using firestore (#6157)
invertase/react-native-firebase: docs/firestore/pagination.md

---
title: Pagination
description: Pagination example using Cloud Firestore.
next: /functions/usage

> previous: /firestore/usage-with-flatlists

Reviewer: In order for this to work, those two pages need to have changes as well to "splice" this page in, since they each have next/prev links.
I think /functions/usage will need it's prev link altered to no longer be /firestore/usage-with-flatlists (it should be /firestore/pagination now)
I think /firestore/usage/with-flatlists will need it's next link changed to firestore/pagination instead of /functions/usage

---

> Pagination using cloud firestore can be done in various ways but here's the very basic way to do it.
> We are going to make the pagination using querying features provided by firestore.

Reviewer: ```suggestion
Pagination using cloud firestore may be done in various ways but here's a basic way to do it using the firestore query features:
```

[`orderBy`, `limit`, `startAfter`]

# Setup state

> First, setup a component which will display the list of data.
> The component will have 2 separate states; `last` and `userData`:

Reviewer: A bit more compact but most importantly the variable is most correctly named lastDocument in the code so reference it as such

```suggestion
First, create a list display component with 2 state items; `lastDocument` and `userData`:
```

```jsx
import React, {useState} from 'react';
import type {Node} from 'react';
import {Text, View, Button, Alert} from 'react-native';

import firestore from '@react-native-firebase/firestore';

const userCollection = firestore().collection('Users');

const App: () => Node = () => {

>   const [lastSnapshot, setLastSnapshot] = useState(null);

Reviewer: ```suggestion
  const [lastDocument, setLastDocument] = useState();
```

  const [userData, setUserData] = useState([]);
};
```

# `LoadData` function

> Next, we'll make a function called `LoadData` that fetches data from `Users` collection.
> This function will be hooked with the `Button`.

Reviewer: ```suggestion
Next, make a function called `LoadData` that fetches data from `Users` collection, and call it when a `Button` is pressed.
```

> If lastSanpshot is not assigned (meaning initial load), the function will fetch from the start.
> After successful fetch from the collection, store the last snapshot data by `setLastSnapshot`.

Reviewer: ```suggestion
If lastDocument is undefined the function will fetch from the star to perform the initial loadt, otherwise it use the lastDocument from the snapshot
After successful fetch from the collection, store the last document we accessed with `setLastDocument`.
```

```jsx
import React, {useState} from 'react';
import type {Node} from 'react';
import {Text, View, Button, Alert} from 'react-native';

import firestore from '@react-native-firebase/firestore';

const userCollection = firestore().collection('Users');

const App: () => Node = () => {

>   const [lastSnapshot, setLastSnapshot] = useState(null);

Reviewer: ```suggestion
  const [lastDocument, setLastDocument] = useState();
```

  const [userData, setUserData] = useState([]);

  function LoadData() {
    console.log('LOAD');

>     if (!lastSnapshot) {
>       userCollection
>         .orderBy('age') //sort the data
>         .limit(3) //get limited amount of data
>         .get()
>         .then(querySnapshot => {
>           setLastSnapshot(querySnapshot.docs[querySnapshot.docs.length - 1]); //set up last snapshot for pagination
>           MakeUserData(querySnapshot.docs); //do what you need to do with the data
>         });
>     } else {
>       userCollection
>         .orderBy('age')
>         .startAfter(lastSnapshot) //fetch data that is placed after the last snapshot that we fetched before.
>         .limit(3)
>         .get()
>         .then(querySnapshot => {
>           setLastSnapshot(querySnapshot.docs[querySnapshot.docs.length - 1]);
>           MakeUserData(querySnapshot.docs);
>         });
>     }

Reviewer: I think this de-duplicates the code?

```suggestion
    let query = userCollection.orderBy('age'); // sort the data
    if (lastSnapshot !== undefined) {
      query = query..startAfter(lastSnapshot); // fetch data following the last document accessed
    }
    query.limit(3) // limit to your page size, 3 is just an example
        .get()
        .then(querySnapshot => {
          setLastDocument(querySnapshot.docs[querySnapshot.docs.length - 1]);
          MakeUserData(querySnapshot.docs);
        });
    }
```

  }

  return (
    <View>
      {userData}
      <Button
        onPress={() => {
          LoadData();
        }}
        title="Load Next"
      />
    </View>
  );
};
```

# `MakeUserData` function

> This is just an example function. You may change this functino into things that you need.

Reviewer: ```suggestion
This is just an example function, alter it to process the data to meet your requirements.
```

In this specific example, it will replace the userData component with the new data fetched.

```js
function MakeUserData(docs) {
  let templist = []; //[...userData] <- use this instead of [] if you want to save the previous data.
  docs.forEach((doc, i) => {
    console.log(doc._data);
    let temp = (
      <View key={i} style={{margin: 10}}>
        <Text>{doc._data.name}</Text>
        <Text>{doc._data.age}</Text>
      </View>
    );
    templist.push(temp);
  });
  setUserData(templist); //replace with the new data
}
```

> Now, everytime you click the button, you will load the data from the `Users` collection.
> If the fetch succesfully completed, you will save the last snapshot and set the userdata component.
> Based on the last snapshot, `LoadData` will fetch the data accordingly.

Reviewer: ```suggestion
Now, every time the button is pressed, `Users` collection data will be fetched one page at a time.
```

# Conclusion

Here's the full example code

```jsx
import React, {useState} from 'react';
import type {Node} from 'react';
import {Text, View, Button, Alert} from 'react-native';

import firestore from '@react-native-firebase/firestore';

const userCollection = firestore().collection('Users');

const App: () => Node = () => {
  const [lastSnapshot, setLastSnapshot] = useState(null);
  const [userData, setUserData] = useState([]);

>   function LoadData() {

Reviewer: If you like the changes above to call it "lastDocument" and to de-duplicate the load code, then the same should be pasted here but I'll wait to see what you think
